### PR TITLE
Use aodnfetcher for jenkins s3 hosted artifacts

### PR DESCRIPTION
**Pull request by anguss00**
Fixes https://github.com/aodn/issues/issues/391 by sourcing artifacts from S3 rather than Jenkins. Will require talend pipeline to archive in S3 only to complete fix.

### CODE REVIEW

File: cookbooks/imos_artifacts/libraries/fetcher_s3.rb [hunk 1]
```diff
@@ -10,51 +10,52 @@
...
-          end
-        end
+        @artifact_filename = artifact_manifest['job']
+        # Assume wheel - TODO fix this assumption, aodnfetcher assumes .war file pattern
+        @artifact_uri = "jenkins://imos-binary/#{job_name}?pattern=^.*\\.whl$"
       end
+    end
 
-      @url = FetcherS3::get_job_url(bucket.url, job_name, buildNumber.to_s, @filename)
+    def download_and_return_path
```
Comment (lwgordonimos) [inline cookbooks/imos_artifacts/libraries/fetcher_s3.rb]: Since the path is in the output JSON, so it would be simpler if this method just returned the parsed JSON document from *here*

File: cookbooks/imos_po/recipes/netcdf_checker.rb [hunk 2]
```diff
@@ -11,6 +11,7 @@
 cchecker_artifact_path = Chef::Recipe::PipelineUtils.get_deployed_artifact_path(node['imos_po']['netcdf_checker']['cchecker_artifact'], node)
 ccplugin_artifact_path = Chef::Recipe::PipelineUtils.get_deployed_artifact_path(node['imos_po']['netcdf_checker']['ccplugin_artifact'], node)
 
+
```
Comment (lwgordonimos) [inline cookbooks/imos_po/recipes/netcdf_checker.rb]: blank line?

File: cookbooks/imos_po/recipes/pipeline.rb [hunk 3]
```diff
@@ -11,6 +11,8 @@
 require 'json'
 require 'pathname'
 
+include_recipe 'imos_po::netcdf_checker'
```
Comment (lwgordonimos) [inline cookbooks/imos_po/recipes/pipeline.rb]: This shouldn't be in the pipeline recipe, since that recipe is for the system netcdf checker, and not related to fetching anyway.

File: private-sample/nodes/po.json [hunk 4]
```diff
@@ -5,6 +5,7 @@
         "role[base]",
         "role[restore]",
         "recipe[ssl]",
+        "recipe[imos_core::aodnfetcher]",
```
Comment (lwgordonimos) [inline private-sample/nodes/po.json]: I don't think this should be a separate standalone recipe, since it's just a dependency of the imos_artifacts cookbook, so that should just pull it in from there if it needs it, not make it the nodes responsibility to make sure it's explicitly installed.

File: cookbooks/imos_artifacts/libraries/fetcher_s3.rb [hunk 5]
```diff
@@ -10,51 +10,52 @@
...
-      buildNumber = keys.map{|item| item.key().split('/')[2].to_i}.max
-
+      @artifact_filename = nil
       if artifact_manifest['filename']
-        @filename = artifact_manifest['filename']
-
+        @artifact_filename = artifact_manifest['filename']
+        artifact_prefix,artifact_suffix = @artifact_filename.split('.')
+        regex_pattern = "^.*#{artifact_prefix}.*\\.#{artifact_suffix}$"
+        @artifact_uri = "jenkins://imos-binary/#{job_name}?pattern=#{regex_pattern}"
```
Comment (lwgordonimos) [inline cookbooks/imos_artifacts/libraries/fetcher_s3.rb]: Should pull out this hard coded bucket to an attribute.

File: cookbooks/imos_artifacts/libraries/fetcher_s3.rb [hunk 6]
```diff
@@ -10,51 +10,52 @@
...
+        # Assume wheel - TODO fix this assumption, aodnfetcher assumes .war file pattern
+        @artifact_uri = "jenkins://imos-binary/#{job_name}?pattern=^.*\\.whl$"
       end
+    end
 
-      @url = FetcherS3::get_job_url(bucket.url, job_name, buildNumber.to_s, @filename)
+    def download_and_return_path
+      artifact_path = nil
+      if find_executable 'aodnfetcher'
+        system_cmd = "cd #{Chef::Config[:file_cache_path]} && aodnfetcher #{@artifact_uri} --cache_dir #{Chef::Config[:file_cache_path]}"
```
Comment (lwgordonimos) [inline cookbooks/imos_artifacts/libraries/fetcher_s3.rb]: Way cleaner to get the fetcher to output a file itself, then you don't need to scrape standard output. The JSON.parse below can just take the file path then.

File: cookbooks/imos_artifacts/libraries/fetcher_s3.rb (hunk 1 above)
Comment (anguss00) [reply cookbooks/imos_artifacts/libraries/fetcher_s3.rb]: Yep, good point. I'll fix that up

**Comment by anguss00** [comment]
@ready to go I think @lwgordonimos

**Comment by anguss00** [comment]
Resource changes up @lwgordonimos to have a gander

File: cookbooks/imos_artifacts/libraries/fetcher_s3.rb [hunk 7]
```diff
@@ -10,51 +10,46 @@
...
+    def download
+      artifact_path = nil
+      outfile = Tempfile.new('aodnfetcher')
+      output_path = outfile.path
+      system_cmd = "cd #{Chef::Config[:file_cache_path]} && aodnfetcher #{@artifact_uri} --outfile #{output_path} --cache_dir #{Chef::Config[:file_cache_path]}"
+      system(system_cmd)
+
+      begin
+        file = File.open(output_path, "rb")
+        contents = file.read
```
Comment (lwgordonimos) [inline cookbooks/imos_artifacts/libraries/fetcher_s3.rb]: Still better to let the JSON library deal with the file object directly!

```
artifact_json = JSON.load(file)
```

File: cookbooks/imos_artifacts/providers/deploy.rb [hunk 8]
```diff
@@ -30,15 +30,19 @@
...
-    Chef::Log.info("Deploying artifact '#{@cached_file_path}' -> '#{dest_file}' -> '#{install_dir}'")
-    ImosArtifacts::Deployer.extract_artifact(
-      @cached_file_path,
-      dest_file,
-      install_dir,
-      new_resource.owner,
-      new_resource.group,
-      new_resource.remove_top_level_directory)
-    new_resource.updated_by_last_action(true)
+    if ! @cached_file_path
```
Comment (lwgordonimos) [inline cookbooks/imos_artifacts/providers/deploy.rb]: This seems unrelated to the change, and looks like "workaround" code still.

File: cookbooks/imos_po/providers/deploy_wheel.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+# Copyright 2019, IMOS
+#
+# All rights reserved - Do Not Redistribute
+#
+# Provider to download pipeline python artifacts
+#
+
+attr_accessor :cached_file_path
+
+action :download do
```
Comment (lwgordonimos) [inline cookbooks/imos_po/providers/deploy_wheel.rb]: 'download' is no longer accurate as the action name, probably should be 'deploy' or 'install'

File: cookbooks/imos_po/resources/deploy_wheel.rb [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+#
+
+actions :download
+default_action :download
+
+attribute :job_name, :kind_of => String, :required => true
+attribute :cached_file_path, :kind_of => String, :required => true
+attribute :venv_dir, :kind_of => String
+attribute :constraints_file, :kind_of => String
+attribute :python, :kind_of => String
```
Comment (lwgordonimos) [inline cookbooks/imos_po/resources/deploy_wheel.rb]: Needs trailing newline

File: private-sample/nodes/po.json (hunk 4 above)
Comment (lwgordonimos) [reply private-sample/nodes/po.json]: ditto on previous omment

File: cookbooks/imos_core/attributes/aodnfetcher.rb [hunk 11]
```diff
@@ -0,0 +1 @@
+default['imos_core']['aodnfetcher']['branch'] = 'master'
```
Comment (lwgordonimos) [inline cookbooks/imos_core/attributes/aodnfetcher.rb]: newline

File: cookbooks/imos_artifacts/recipes/default.rb [hunk 12]
```diff
@@ -12,5 +12,7 @@
 # Needed to build ruby gems
 include_recipe "build-essential"
 
+include_recipe 'imos_core::aodnfetcher'
```
Comment (lwgordonimos) [inline cookbooks/imos_artifacts/recipes/default.rb]: Just inline this, so we don't end up with more recipes just for dependencies

**Comment by anguss00** [comment]
@lwgordonimos 
giphy (18)
